contributors/design-proposals/hugepages.md

```diff
@@ -0,0 +1,288 @@
+reserved.
+
+```
+apiVersion: v1
+kind: Node
+metadata:
+  name: node1
+...
+status:
+  capacity:
+    memory: 10Gi
+    hugepages.2Mi: 512
+  allocatable:
+    memory: 9Gi
+    hugepages.2Mi: 512
+...  
+```
+
+## Pod Specification
+
+A pod must make a request to consume pre-allocated huge pages using the resource
+`hugepages.<hugepagesize>` whose quantity is a positive number of pages. The
+request and limit for `hugepages.<hugepagesize>` must match until such time the
+node provides more reliability in handling overcommit (potentially via
+eviction).  An application that consumes the `hugepages.<hugepagesize>` resource
+is at minimum in the `Burstable` QoS class.
+
+If a pod consumes huge pages via `shmget`, it must run with a supplemental group
+that matches `/proc/sys/vm/hugetlb_shm_group` on the node.  Configuration of
+this group is outside the scope of this specification.
+
+In order to consume huge pages backed by the `hugetlbfs` filesystem inside the
+specified container in the pod, it is helpful to understand the set of mount
+options used with `hugetlbfs`.  For more details, see "Using Huge Pages" here:
+https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt
+
+```
+mount -t hugetlbfs \
+	-o uid=<value>,gid=<value>,mode=<value>,pagesize=<value>,size=<value>,\
+	min_size=<value>,nr_inodes=<value> none /mnt/huge
+```
+
+The proposal recommends extending the existing `EmptyDirVolumeSource` to satisfy
+this use case.  A new `medium=HugePages` option would be supported.  A new
+optional field `pageSize` would be required when using this medium.  The
+`pageSize` option aligns with the `pagesize` argument in the above mount.  It is
+used to specify the huge page size and pool that is used for accounting.  To
+write into this volume, the pod must make a request for huge pages that matches
+this `pageSize`, otherwise its usage will be limited to 0 bytes.  The existing
+`sizeLimit` option for `emptyDir` would restrict usage to the minimum value
+specified between `sizeLimit` and the sum of huge page limits of all containers
+in a pod for the specified `pageSize`. This keeps the behavior consistent with
+memory backed `emptyDir` volumes whose usage is ultimately constrained by the
+pod cgroup sandbox memory settings.  The `min_size` option is omitted as its not
+necessary.  The `nr_inodes` mount option is omitted at this time in the same
+manner it is omitted with `medium=Memory` when using `tmpfs`.
+
+The following is a sample pod that is limited to 512 huge pages of size 2MiB. It
+can consume those pages using `shmget()` or via `mmap()` with the specified
+volume.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+...
+    volumeMounts:
+    - mountPath: /hugepages
+      name: hugepage
+    resources:
+      requests:
+        hugepages.2Mi: "512"
+      limits:
+        hugepages.2Mi: "512"
+  volumes:
+  - name: hugepage
+    emptyDir:
+      medium: HugePages
+      pageSize: "2Mi"
```

Comment:
Yes - It's probably good to tie the `EmptyDir` to the request hugepages. Otherwise there is also the option of abusing the mechanism, by requesting one size, but then using a different `pageSize`.

Quite unlikely, but still possible is that different hugepage sizes are requested, thus the volume needs a reference to the hugepage size. Just saying that a volume can not infer the page size automatically if there is more than one given. This in turn would mean that the volume should reference the resource (thus `hugepages.2Mi`in the example above)